### ISSUE TITLE
Introduce caching for adviser so you can prevent DDoS + be more efficient with resources

### ISSUE THREAD

**Issue by tumido** [thoth-station/thoth-application#1025]
**Is your feature request related to a problem? Please describe.**
Thoth team seems to be always constraint on resources, one time it's PSI folks complaining about the number of events, other time it's operate first networking not able to keep up. If AICoE-CI wants to scale with OperateFirst and expand on `thamos` usage, I think it would be smart to think about de-duplication of repeated and "the same" workloads. 

**Describe the solution you'd like**
Introduce a response caching mechanism for same stack resolutions and request aggregation on simultaneous resolution requests of the same stack from different users. 

Imagine a workshop scenario, when 50+ different people are at the same time requesting a resolution of the same stack. Because it's a workshop, the stack is the same and since they are 50+ different people their user id differs. In current situation this would results with many workflows being spawned - 1 for every request of every user for every notebook for example. These requests would usually happen at the ~same time, since the workshops are coordinated.

I know that Thoth wants to deliver as up-to-date recommendations as possible and basically immediately after any new package version is released to provide recommendation upon that. To mitigate the possibility of recommending a outdated stack, I'd suggest you to consider short cache duration 10-30mins and dynamic cache invalidation - if the new package is part of the resolved stack, the cache is invalidated.

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

**Comment by fridex** [comment]
hi @tumido. We have caching for each request to user-api available. You can try it using:

```
thamos advise --no-wait
thamos advise --no-wait
```

This way, you should be able to receive the same adviser id unless `--force` is provided. We have a bug in stage/prod deployments where cache is not properly hit when users submit software stack with static source code analysis (see https://github.com/thoth-station/user-api/issues/1324).

If the issue is related to some specific case, please let us know a reproducer we would be happy to help.

**Comment by fridex** [comment]
Actually, I think the cache won't be hit if you use the same configuration but provide a different origin (e.g. your fork). We track the use of the stacks, so this could be something to consider to be improved. Can this be something you experience? E.g. you create a fork of the main repo to your GH and do advise on it for each fork in a workshop?

**Comment by tumido** [comment]
I'm thinking of more in a context of a workshop right now:

50+ people fork a repo, pull it into Jupyter Lab, then they try to resolve dependencies via the Jupyter Lab plugin from a notebook, all at the same time. As far as I know this will not be mapped under the same advised ID and each of the 50 people would get a different ID and a different ID for every notebook, won't they?

**Comment by fridex** [comment]
> I'm thinking of more in a context of a workshop right now:
> 
> 50+ people fork a repo, pull it into Jupyter Lab, then they try to resolve dependencies via the Jupyter Lab plugin from a notebook, all at the same time. As far as I know this will not be mapped under the same advised ID and each of the 50 people would get a different ID and a different ID for every notebook, won't they?

They should receive the same adviser id from user-api side unless `force` to the request is passed:

https://github.com/thoth-station/user-api/blob/6fad0cedd76143d98013d27917e0382ba8671ea0/openapi/openapi.yaml#L578-L585

One can verify if the adviser id provided by user-api is cached by checking `cached` in the response (might be too low level to show it in client tools):

https://github.com/thoth-station/user-api/blob/6fad0cedd76143d98013d27917e0382ba8671ea0/openapi/openapi.yaml#L1508-L1511

This is done on user-api before scheduling any workload so this should be transparent to users and agnostic to the client tool used (jypyuterlab extension, thamos, ...).

The use case when this will not work is different origin as it will not cause cache hit. If you fork the same repo and I fork the same repo, we do an advise on them, the cache will not be hit properly as the origin differentiates our requests:

https://github.com/thoth-station/user-api/blob/6fad0cedd76143d98013d27917e0382ba8671ea0/openapi/openapi.yaml#L535-L542

We could make this work in a way the origin field will not differentiate cache and should solve the problem I guess.

**Comment by tumido** [comment]
> We could make this work in a way the origin field will not differentiate cache and should solve the problem I guess.

That sounds like a viable solution. Just an idea (without knowing the technical details, so please bear with me): what about you use the Pipfile hash in here instead? Like the same way it's used in the lock file to confirm integrity. That should identify unique stacks better than the origin.

**Comment by fridex** [comment]
> That sounds like a viable solution. Just an idea (without knowing the technical details, so please bear with me): 

We are open to all ideas so feel free to suggest anything without any hesitation. :+1: Your input is completely valid.

> what about you use the Pipfile hash in here instead? Like the same way it's used in the lock file to confirm integrity. That should identify unique stacks better than the origin.

Pipfile hash is specific to Pipfile Python requirements. As we moved one step up and considered also hardware, software, API use we had to compound all these to compute the hash for a request to the recommendation engine:

https://github.com/thoth-station/user-api/blob/6fad0cedd76143d98013d27917e0382ba8671ea0/thoth/user_api/api_v1.py#L352-L368

An example of a wrong cache hit could be a recommendation specific to tensorflow when a user uses tensorflow API that is available only for tensorflow>=2.0 and asks for an advise. Let's say we suggest to use tensorflow==2.3.0. The next step by a user is a change in tensorflow API used and the user starts to use symbols available in tensorflow>=2.4.0. Based on the library usage, we cannot suggest using tensorflow==2.3.0 from cache again on the next advise, but we need to trigger a new advise that will take into account the new API used even though there were no changes in the requirement files by the user. The same could be applied to hardware, software, recommendation types, and other options that are accepted by the recommendation engine to compute advise.

**Comment by tumido** [comment]
ah, that totally make sense, yeah. I didn't considered that at all.

So if I understand this correctly, this whole dict of properties is used to calculate a hash that identifies the cache entry, correct?
> https://github.com/thoth-station/user-api/blob/6fad0cedd76143d98013d27917e0382ba8671ea0/thoth/user_api/api_v1.py#L352-L368

Let me try to differentiate the "inner" (important to the recommendation) and "outer" (important to identify the requester) factors of this cache entry ID:

| Property                 | Significance                                                                                              | Outcome                                                    |
| ------------------------ | --------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------- |
| `project`                | A change in this field changes the response                                                               | Can't be omitted                                           |
| `count`                  | A change in this field changes the response                                                               | Can't be omitted                                           |
| `limit`                  | A change in this field changes the response                                                               | Can't be omitted                                           |
| `library_usage`          | A change in this field changes the response                                                               | Can't be omitted                                           |
| `recommendation_type`    | A change in this field changes the response                                                               | Can't be omitted                                           |
| `origin`                 | No importance for the response content                                                                    | Possible to remove                                         |
| `source_type`            | This is already covered by the project, isn't it? A different source type would be reflected in `project` | Possible to remove (May be already redundant)                                     |
| `dev`                    | A change in this field changes the response                                                               | Can't be omitted                                           |
| `debug`                  | No idea if this changes the response content or not.                                                      | If debug is set, it may be omitted from caching every time |
| `github_event_type`      | No importance for the response content                                                                    | Possible to remove                                         |
| `github_check_run_id`    | No importance for the response content                                                                    | Possible to remove                                         |
| `github_installation_id` | No importance for the response content                                                                    | Possible to remove                                         |
| `github_base_repo_url`   | No importance for the response content                                                                    | Possible to remove                                         |

Please correct me if I got any of it wrong.

I think limiting the factors that defined the cache ID to the bare minimum - a.k.a those factors that actually makes the difference on the recommendation response only... (easier to said than done, right)

**Comment by fridex** [comment]
> ah, that totally make sense, yeah. I didn't considered that at all.
> 
> So if I understand this correctly, this whole dict of properties is used to calculate a hash that identifies the cache entry, correct?
> > https://github.com/thoth-station/user-api/blob/6fad0cedd76143d98013d27917e0382ba8671ea0/thoth/user_api/api_v1.py#L352-L368

Yes, that's correct :+1: 

> Let me try to differentiate the "inner" (important to the recommendation) and "outer" (important to identify the requester) factors of this cache entry ID:

Wow, nice write-up! :+1: 

> `source_type`	This is already covered by the project, isn't it? A different source type would be reflected in `project`	Possible to remove (May be already redundant)

This property states "submitter" of the request - it can be s2i, cli tool or similar. We keep it for tracking usage and this is not captured in the `project` abstraction.

> `dev`	A change in this field changes the response	Can't be omitted
> `debug`	No idea if this changes the response content or not.	If debug is set, it may be omitted from caching every time

I think this is not true. Running an advise in debug mode results in a slower adviser run as it needs to print debug messages to a slow terminal. As we produce results given the time allocated, running adviser in a debug mode and without debug mode can result in different results (adviser will be slower, will not browse as much of the state space as without the debug flag).

> `github_event_type`	No importance for the response content	Possible to remove
> `github_check_run_id`	No importance for the response content	Possible to remove
> `github_installation_id`	No importance for the response content	Possible to remove
> `github_base_repo_url`	No importance for the response content	Possible to remove

To handle these properly with cache, additional business logic will be needed. These are used to send results back to github once an advise is done. If a user clones a repo, does `thamos advise` and then goes to GitHub and triggers our integration on GH, the result will not be properly sent back to GitHub as internal workflow handling GitHub events will not be triggered (user-api will respond with the cached results).

GH handling, as well as "submitter" handling, can be indeed optimized when it comes to cache usage. It will require additional logic to be implemented to work properly though. I'm on your side with optimizing this, but given the lack of handling logic of these corner cases, we are not able to remove these entries right away without affecting users or without making us work properly with services we integrate with.

> I think limiting the factors that defined the cache ID to the bare minimum - a.k.a those factors that actually makes the difference on the recommendation response only... (easier to said than done, right)

+1

**Comment by pacospace** [comment]
can we consider this issue solved after the changes in user-api?

**Comment by fridex** [comment]
Yes, I think we can close this one and have #1112 as issue with references. Thanks to everyone involved.

/close
